In [1]:
import numpy as np

import glob
import re
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
import scipy.stats.stats as stats

#from sklearn.decomposition import PCA

In [2]:
import zipfile

# read the dataset using the compression zip
all_data = pd.read_csv('all_data.zip',compression='zip')

In [3]:
all_data

PR_number                                            Comment  \
0       12027_rev1_d                                                NaN   
1       12050_rev1_d                                                NaN   
2       12050_rev2_d                              Uploaded patch set 2.   
3       12050_rev2_d  Patch Set 2: VerifiedMatt/Dustin/Mike:I have r...   
4       12050_rev2_d  Patch Set 2: I would prefer that you didn't su...   
...              ...                                                ...   
422615  85948_rev2_d  Change has been successfully cherry-picked as ...   
422616  85957_rev1_d                              Uploaded patch set 1.   
422617  85957_rev1_d                         Patch Set 1: Code-Review+2   
422618  85957_rev1_d                            Patch Set 1: Verified+1   
422619  85957_rev2_d  Change has been successfully cherry-picked as ...   

               Author                           Date  \
0                 NaN                            NaN   
1                 NaN                            NaN   
2        Jeffrey Roth  2012-01-04 21:45:42.382000000   
3        Jeffrey Roth  2012-01-04 21:51:47.408000000   
4       Kenny Maynard  2012-01-04 22:01:33.920000000   
...               ...                            ...   
422615   Emely Wilson  2017-11-23 10:51:00.946000000   
422616   Emely Wilson  2017-11-23 16:01:28.983000000   
422617   Alissa Huang  2017-11-23 16:04:40.102000000   
422618   Emely Wilson  2017-11-23 16:05:06.479000000   
422619   Emely Wilson  2017-11-23 16:05:08.701000000   

                                           Commit_message  \
0                                                     NaN   
1                                                     NaN   
2       Refactoring changes summary:Membase has been r...   
3       Refactoring changes summary:Membase has been r...   
4       Refactoring changes summary:Membase has been r...   
...                                                   ...   
422615                                                NaN   
422616                                                NaN   
422617                                                NaN   
422618                                                NaN   
422619                                                NaN   

                                             Descriptions     status  \
0                                                     NaN  ABANDONED   
1                                                     NaN     MERGED   
2       Potential Refactoring of Java Client to couchb...     MERGED   
3       Potential Refactoring of Java Client to couchb...     MERGED   
4       Potential Refactoring of Java Client to couchb...     MERGED   
...                                                   ...        ...   
422615          CBQE-4333: Fix the data population issues     MERGED   
422616  CBQE-4333: make eventing dataset tc's work fro...     MERGED   
422617  CBQE-4333: make eventing dataset tc's work fro...     MERGED   
422618  CBQE-4333: make eventing dataset tc's work fro...     MERGED   
422619  CBQE-4333: make eventing dataset tc's work fro...     MERGED   

          main_author  revision_number  \
0       Emerson Nolan                1   
1       Emerson Nolan                1   
2       Emerson Nolan                2   
3       Emerson Nolan                2   
4       Emerson Nolan                2   
...               ...              ...   
422615   Emely Wilson                2   
422616   Emely Wilson                1   
422617   Emely Wilson                1   
422618   Emely Wilson                1   
422619   Emely Wilson                2   

                                            url  
0       http://review.couchbase.org/#/c/12027/1  
1       http://review.couchbase.org/#/c/12050/1  
2       http://review.couchbase.org/#/c/12050/2  
3       http://review.couchbase.org/#/c/12050/2  
4       http://review.couchbase.org/#/c/12050/2  
...                               

Convert Status to 0 if abandoned and 1 if merged

In [4]:
all_data['status'] = np.where((all_data['status'] == 'MERGED') ,True,False)*1

In [5]:
all_data

PR_number                                            Comment  \
0       12027_rev1_d                                                NaN   
1       12050_rev1_d                                                NaN   
2       12050_rev2_d                              Uploaded patch set 2.   
3       12050_rev2_d  Patch Set 2: VerifiedMatt/Dustin/Mike:I have r...   
4       12050_rev2_d  Patch Set 2: I would prefer that you didn't su...   
...              ...                                                ...   
422615  85948_rev2_d  Change has been successfully cherry-picked as ...   
422616  85957_rev1_d                              Uploaded patch set 1.   
422617  85957_rev1_d                         Patch Set 1: Code-Review+2   
422618  85957_rev1_d                            Patch Set 1: Verified+1   
422619  85957_rev2_d  Change has been successfully cherry-picked as ...   

               Author                           Date  \
0                 NaN                            NaN   
1                 NaN                            NaN   
2        Jeffrey Roth  2012-01-04 21:45:42.382000000   
3        Jeffrey Roth  2012-01-04 21:51:47.408000000   
4       Kenny Maynard  2012-01-04 22:01:33.920000000   
...               ...                            ...   
422615   Emely Wilson  2017-11-23 10:51:00.946000000   
422616   Emely Wilson  2017-11-23 16:01:28.983000000   
422617   Alissa Huang  2017-11-23 16:04:40.102000000   
422618   Emely Wilson  2017-11-23 16:05:06.479000000   
422619   Emely Wilson  2017-11-23 16:05:08.701000000   

                                           Commit_message  \
0                                                     NaN   
1                                                     NaN   
2       Refactoring changes summary:Membase has been r...   
3       Refactoring changes summary:Membase has been r...   
4       Refactoring changes summary:Membase has been r...   
...                                                   ...   
422615                                                NaN   
422616                                                NaN   
422617                                                NaN   
422618                                                NaN   
422619                                                NaN   

                                             Descriptions  status  \
0                                                     NaN       0   
1                                                     NaN       1   
2       Potential Refactoring of Java Client to couchb...       1   
3       Potential Refactoring of Java Client to couchb...       1   
4       Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
422615          CBQE-4333: Fix the data population issues       1   
422616  CBQE-4333: make eventing dataset tc's work fro...       1   
422617  CBQE-4333: make eventing dataset tc's work fro...       1   
422618  CBQE-4333: make eventing dataset tc's work fro...       1   
422619  CBQE-4333: make eventing dataset tc's work fro...       1   

          main_author  revision_number  \
0       Emerson Nolan                1   
1       Emerson Nolan                1   
2       Emerson Nolan                2   
3       Emerson Nolan                2   
4       Emerson Nolan                2   
...               ...              ...   
422615   Emely Wilson                2   
422616   Emely Wilson                1   
422617   Emely Wilson                1   
422618   Emely Wilson                1   
422619   Emely Wilson                2   

                                            url  
0       http://review.couchbase.org/#/c/12027/1  
1       http://review.couchbase.org/#/c/12050/1  
2       http://review.couchbase.org/#/c/12050/2  
3       http://review.couchbase.org/#/c/12050/2  
4       http://review.couchbase.org/#/c/12050/2  
...                                         ...  
422615  http://revie

In [6]:
all_data['PR_number'].isna().sum()

0

In [7]:
all_data.isna().sum()

PR_number               0
Comment             13819
Author              13838
Date                13836
Commit_message     164978
Descriptions        13866
status                  0
main_author             5
revision_number         0
url                     0
dtype: int64

Drop all rows with nan comments, Author and Dates

In [8]:
all_data = all_data.dropna(subset=['Comment'])
all_data = all_data.dropna(subset=['Author'])
all_data = all_data.dropna(subset=['Date'])
all_data = all_data.dropna(subset=['main_author'])

In [9]:
all_data

PR_number                                            Comment  \
2       12050_rev2_d                              Uploaded patch set 2.   
3       12050_rev2_d  Patch Set 2: VerifiedMatt/Dustin/Mike:I have r...   
4       12050_rev2_d  Patch Set 2: I would prefer that you didn't su...   
5       12050_rev2_d  Patch Set 2:You also need to fix the commit me...   
6       12050_rev2_d  Patch Set 2:I don't know that changing the pac...   
...              ...                                                ...   
422615  85948_rev2_d  Change has been successfully cherry-picked as ...   
422616  85957_rev1_d                              Uploaded patch set 1.   
422617  85957_rev1_d                         Patch Set 1: Code-Review+2   
422618  85957_rev1_d                            Patch Set 1: Verified+1   
422619  85957_rev2_d  Change has been successfully cherry-picked as ...   

               Author                           Date  \
2        Jeffrey Roth  2012-01-04 21:45:42.382000000   
3        Jeffrey Roth  2012-01-04 21:51:47.408000000   
4       Kenny Maynard  2012-01-04 22:01:33.920000000   
5       Kenny Maynard  2012-01-04 22:02:19.848000000   
6       Andreas Short  2012-01-05 00:24:39.162000000   
...               ...                            ...   
422615   Emely Wilson  2017-11-23 10:51:00.946000000   
422616   Emely Wilson  2017-11-23 16:01:28.983000000   
422617   Alissa Huang  2017-11-23 16:04:40.102000000   
422618   Emely Wilson  2017-11-23 16:05:06.479000000   
422619   Emely Wilson  2017-11-23 16:05:08.701000000   

                                           Commit_message  \
2       Refactoring changes summary:Membase has been r...   
3       Refactoring changes summary:Membase has been r...   
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
6       Refactoring changes summary:Membase has been r...   
...                                                   ...   
422615                                                NaN   
422616                                                NaN   
422617                                                NaN   
422618                                                NaN   
422619                                                NaN   

                                             Descriptions  status  \
2       Potential Refactoring of Java Client to couchb...       1   
3       Potential Refactoring of Java Client to couchb...       1   
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
6       Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
422615          CBQE-4333: Fix the data population issues       1   
422616  CBQE-4333: make eventing dataset tc's work fro...       1   
422617  CBQE-4333: make eventing dataset tc's work fro...       1   
422618  CBQE-4333: make eventing dataset tc's work fro...       1   
422619  CBQE-4333: make eventing dataset tc's work fro...       1   

          main_author  revision_number  \
2       Emerson Nolan                2   
3       Emerson Nolan                2   
4       Emerson Nolan                2   
5       Emerson Nolan                2   
6       Emerson Nolan                2   
...               ...              ...   
422615   Emely Wilson                2   
422616   Emely Wilson                1   
422617   Emely Wilson                1   
422618   Emely Wilson                1   
422619   Emely Wilson                2   

                                            url  
2       http://review.couchbase.org/#/c/12050/2  
3       http://review.couchbase.org/#/c/12050/2  
4       http://review.couchbase.org/#/c/12050/2  
5       http://review.couchbase.org/#/c/12050/2  
6       http://review.couchbase.org/#/c/12050/2  
...                                         ...  
422615  http://revie

Add a column with only the Pr-number without the patch version

In [10]:
all_data['PR_number_no_patch_number'] = all_data['PR_number'].str[:5]

In [11]:
all_data

PR_number                                            Comment  \
2       12050_rev2_d                              Uploaded patch set 2.   
3       12050_rev2_d  Patch Set 2: VerifiedMatt/Dustin/Mike:I have r...   
4       12050_rev2_d  Patch Set 2: I would prefer that you didn't su...   
5       12050_rev2_d  Patch Set 2:You also need to fix the commit me...   
6       12050_rev2_d  Patch Set 2:I don't know that changing the pac...   
...              ...                                                ...   
422615  85948_rev2_d  Change has been successfully cherry-picked as ...   
422616  85957_rev1_d                              Uploaded patch set 1.   
422617  85957_rev1_d                         Patch Set 1: Code-Review+2   
422618  85957_rev1_d                            Patch Set 1: Verified+1   
422619  85957_rev2_d  Change has been successfully cherry-picked as ...   

               Author                           Date  \
2        Jeffrey Roth  2012-01-04 21:45:42.382000000   
3        Jeffrey Roth  2012-01-04 21:51:47.408000000   
4       Kenny Maynard  2012-01-04 22:01:33.920000000   
5       Kenny Maynard  2012-01-04 22:02:19.848000000   
6       Andreas Short  2012-01-05 00:24:39.162000000   
...               ...                            ...   
422615   Emely Wilson  2017-11-23 10:51:00.946000000   
422616   Emely Wilson  2017-11-23 16:01:28.983000000   
422617   Alissa Huang  2017-11-23 16:04:40.102000000   
422618   Emely Wilson  2017-11-23 16:05:06.479000000   
422619   Emely Wilson  2017-11-23 16:05:08.701000000   

                                           Commit_message  \
2       Refactoring changes summary:Membase has been r...   
3       Refactoring changes summary:Membase has been r...   
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
6       Refactoring changes summary:Membase has been r...   
...                                                   ...   
422615                                                NaN   
422616                                                NaN   
422617                                                NaN   
422618                                                NaN   
422619                                                NaN   

                                             Descriptions  status  \
2       Potential Refactoring of Java Client to couchb...       1   
3       Potential Refactoring of Java Client to couchb...       1   
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
6       Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
422615          CBQE-4333: Fix the data population issues       1   
422616  CBQE-4333: make eventing dataset tc's work fro...       1   
422617  CBQE-4333: make eventing dataset tc's work fro...       1   
422618  CBQE-4333: make eventing dataset tc's work fro...       1   
422619  CBQE-4333: make eventing dataset tc's work fro...       1   

          main_author  revision_number  \
2       Emerson Nolan                2   
3       Emerson Nolan                2   
4       Emerson Nolan                2   
5       Emerson Nolan                2   
6       Emerson Nolan                2   
...               ...              ...   
422615   Emely Wilson                2   
422616   Emely Wilson                1   
422617   Emely Wilson                1   
422618   Emely Wilson                1   
422619   Emely Wilson                2   

                                            url PR_number_no_patch_number  
2       http://review.couchbase.org/#/c/12050/2                     12050  
3       http://review.couchbase.org/#/c/12050/2                     12050  
4       http://review.couchbase.org/#/c/12050/2                     12050  
5       http://review.couchbase.org/#/c/12050/2                   

Deleting all the comments from bots.

In [12]:
bot_comments = ['Uploaded patch set','Added to reviewer','Verified','Change has been successfully','Your change could not be merged due to a path conflict.','Code-Review','(1 comment)'
                ,'Logs','log','Cherry Picked', 'was rebased','Cherry Picked from branch master.','cherry picked',
                'Cherry Picked','comments','Published edit on patch set']
bot_comments_pattern = '|'.join(bot_comments)


In [13]:
all_data.Comment.str.contains(bot_comments_pattern).sum()

C:\Users\GNR\AppData\Local\Temp\ipykernel_588\1003850507.py:1: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  all_data.Comment.str.contains(bot_comments_pattern).sum()


287332

We have 287332 comments from bots

And now we are going to delete them.

In [14]:
all_data = all_data[all_data.Comment.str.contains(bot_comments_pattern) == False]

C:\Users\GNR\AppData\Local\Temp\ipykernel_588\3367746845.py:1: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  all_data = all_data[all_data.Comment.str.contains(bot_comments_pattern) == False]


In [15]:
all_data

PR_number                                            Comment  \
4       12050_rev2_d  Patch Set 2: I would prefer that you didn't su...   
5       12050_rev2_d  Patch Set 2:You also need to fix the commit me...   
6       12050_rev2_d  Patch Set 2:I don't know that changing the pac...   
7       12050_rev2_d  Patch Set 2:I don't know that changing the pac...   
8       12050_rev2_d  Patch Set 2: No scoreThe only reason I brought...   
...              ...                                                ...   
422502  85713_rev2_d  Created a revert of this change as I4c00a1f817...   
422544  85827_rev3_d  MISMATCHED INLINE COMMENTLine:44, pytests/even...   
422560  85859_rev1_d  Assignee added: Nelson Emerson Dukekson <Nelso...   
422561  85859_rev1_d  Assignee deleted: Nelson Emerson Dukekson <Nel...   
422611  85939_rev1_d                              Abandonedtesting only   

                Author                           Date  \
4        Kenny Maynard  2012-01-04 22:01:33.920000000   
5        Kenny Maynard  2012-01-04 22:02:19.848000000   
6        Andreas Short  2012-01-05 00:24:39.162000000   
7        Andreas Short  2012-01-05 00:24:39.162000000   
8        Kenny Maynard  2012-01-05 18:45:06.860000000   
...                ...                            ...   
422502    Alissa Huang  2017-11-19 04:56:20.273000000   
422544    Emely Wilson  2017-11-21 11:03:55.013000000   
422560  Jakobe Johnson  2017-11-21 21:26:16.466000000   
422561  Jakobe Johnson  2017-11-21 21:27:16.130000000   
422611    Johan Savage  2017-11-23 11:45:34.197000000   

                                           Commit_message  \
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
6       Refactoring changes summary:Membase has been r...   
7       Refactoring changes summary:Membase has been r...   
8       Refactoring changes summary:Membase has been r...   
...                                                   ...   
422502                                                NaN   
422544                                                NaN   
422560                       Fix one problem with STEP 1.   
422561                       Fix one problem with STEP 1.   
422611                                                NaN   

                                             Descriptions  status  \
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
6       Potential Refactoring of Java Client to couchb...       1   
7       Potential Refactoring of Java Client to couchb...       1   
8       Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
422502  CBQE-4388 set ulimit to 102400 in centos serve...       1   
422544        CBQE-4333: Add eventing negative test cases       1   
422560             CBQE-4390:Fix one problem with STEP 1.       1   
422561             CBQE-4390:Fix one problem with STEP 1.       1   
422611                         Test change, do not submit       0   

             main_author  revision_number  \
4          Emerson Nolan                2   
5          Emerson Nolan                2   
6          Emerson Nolan                2   
7          Emerson Nolan                2   
8          Emerson Nolan                2   
...                  ...              ...   
422502  Frederick Booker                2   
422544      Emely Wilson                3   
422560    Jakobe Johnson                1   
422561    Jakobe Johnson                1   
422611      Johan Savage                1   

                                            url PR_number_no_patch_number  
4       http://review.couchbase.org/#/c/12050/2                     12050  
5       http://review.couchbase.org/#/c/12050/2                     12050  
6       http://review.couchbase.org/#/c/12050/2                     12050  
7       http://rev

I noticed that only the comments that start with 'Patch Set {number of the patch} are useful for our analysis so I am going to keep only those comments.

In [16]:
useful_data = all_data[all_data['Comment'].str[:9].str.contains('Patch Set') == True]

In [17]:
useful_data

PR_number                                            Comment  \
4       12050_rev2_d  Patch Set 2: I would prefer that you didn't su...   
5       12050_rev2_d  Patch Set 2:You also need to fix the commit me...   
6       12050_rev2_d  Patch Set 2:I don't know that changing the pac...   
7       12050_rev2_d  Patch Set 2:I don't know that changing the pac...   
8       12050_rev2_d  Patch Set 2: No scoreThe only reason I brought...   
...              ...                                                ...   
420134  80604_rev2_d           Patch Set 2:Got it - will remember that.   
420185  80992_rev1_d  Patch Set 1:MemoryQuota may set in param with ...   
420805  82094_rev1_d  Patch Set 1:This change has already been revie...   
421737  84162_rev1_d  Patch Set 1:test suite lww in entbackup restor...   
422442  85613_rev1_d  Patch Set 1:Need to verify it on both, centos ...   

                  Author                           Date  \
4          Kenny Maynard  2012-01-04 22:01:33.920000000   
5          Kenny Maynard  2012-01-04 22:02:19.848000000   
6          Andreas Short  2012-01-05 00:24:39.162000000   
7          Andreas Short  2012-01-05 00:24:39.162000000   
8          Kenny Maynard  2012-01-05 18:45:06.860000000   
...                  ...                            ...   
420134        Ellie Kidd  2017-07-13 18:41:21.038000000   
420185  Frederick Booker  2017-07-18 15:43:30.472000000   
420805    Monique Barker  2017-08-09 02:25:02.503000000   
421737  Frederick Booker  2017-10-09 19:52:13.747000000   
422442  Frederick Booker  2017-11-17 18:42:57.845000000   

                                           Commit_message  \
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
6       Refactoring changes summary:Membase has been r...   
7       Refactoring changes summary:Membase has been r...   
8       Refactoring changes summary:Membase has been r...   
...                                                   ...   
420134                                                NaN   
420185                                                NaN   
420805                                                NaN   
421737                                                NaN   
422442                                                NaN   

                                             Descriptions  status  \
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
6       Potential Refactoring of Java Client to couchb...       1   
7       Potential Refactoring of Java Client to couchb...       1   
8       Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
420134  CBQE-3067: Use better FTS level compaction con...       1   
420185  CBQE-4209: Added the cbas installation and ser...       1   
420805            Adding subquery tests to watson branch.       1   
421737  CBQE-1898 lww tests are covered in py-entbacku...       1   
422442  CBQE-4378 String manipulation in bash incorrec...       1   

             main_author  revision_number  \
4          Emerson Nolan                2   
5          Emerson Nolan                2   
6          Emerson Nolan                2   
7          Emerson Nolan                2   
8          Emerson Nolan                2   
...                  ...              ...   
420134        Ellie Kidd                2   
420185       Alfred Park                1   
420805        Dylan Paul                1   
421737  Frederick Booker                1   
422442    Angelina Mayer                1   

                                            url PR_number_no_patch_number  
4       http://review.couchbase.org/#/c/12050/2                     12050  
5       http://review.couchbase.org/#/c/12050/2                     12050  
6       http://review.couchbase.org/#/c/12050/2                     12

Now we are going to remove the Patch set number from every comment.

In [18]:
useful_data['Comment'] = all_data['Comment'].str[12:]

C:\Users\GNR\AppData\Local\Temp\ipykernel_588\13972313.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  useful_data['Comment'] = all_data['Comment'].str[12:]


In [19]:
bot_comments = ['Patch Set','patch set']
bot_comments_pattern = '|'.join(bot_comments)
all_data = all_data[all_data.Comment.str.contains(bot_comments_pattern) == False]
all_data.Comment.str.contains(bot_comments_pattern).sum()

0

Find the length of every comment(METRICS)

In [20]:
useful_data['comment_length'] = useful_data['Comment'].str.len()

C:\Users\GNR\AppData\Local\Temp\ipykernel_588\2463225354.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  useful_data['comment_length'] = useful_data['Comment'].str.len()


Removing all the comments with more than 500 characters.(METRICS)


In [21]:
useful_data = useful_data[useful_data['comment_length'] <= 500]

In [22]:
useful_data

PR_number                                            Comment  \
4       12050_rev2_d   I would prefer that you didn't submit thisI w...   
5       12050_rev2_d  You also need to fix the commit message so it ...   
9       12050_rev2_d   I would prefer that you didn't submit this-1 ...   
10      12050_rev2_d  I don't have a huge opinion on it, really.  My...   
12      12050_rev3_d  I think Rags is working on this today too, so ...   
...              ...                                                ...   
420134  80604_rev2_d                       Got it - will remember that.   
420185  80992_rev1_d  MemoryQuota may set in param with default valu...   
420805  82094_rev1_d  This change has already been reviewed by dev f...   
421737  84162_rev1_d  test suite lww in entbackup restore is also re...   
422442  85613_rev1_d      Need to verify it on both, centos and windows   

                  Author                           Date  \
4          Kenny Maynard  2012-01-04 22:01:33.920000000   
5          Kenny Maynard  2012-01-04 22:02:19.848000000   
9          Kenny Maynard  2012-01-05 22:42:21.672000000   
10             Noah Dean  2012-01-06 03:50:33.697000000   
12         Andreas Short  2012-01-08 21:36:57.811000000   
...                  ...                            ...   
420134        Ellie Kidd  2017-07-13 18:41:21.038000000   
420185  Frederick Booker  2017-07-18 15:43:30.472000000   
420805    Monique Barker  2017-08-09 02:25:02.503000000   
421737  Frederick Booker  2017-10-09 19:52:13.747000000   
422442  Frederick Booker  2017-11-17 18:42:57.845000000   

                                           Commit_message  \
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
9       Refactoring changes summary:Membase has been r...   
10      Refactoring changes summary:Membase has been r...   
12      Refactoring changes summary:Membase has been r...   
...                                                   ...   
420134                                                NaN   
420185                                                NaN   
420805                                                NaN   
421737                                                NaN   
422442                                                NaN   

                                             Descriptions  status  \
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
9       Potential Refactoring of Java Client to couchb...       1   
10      Potential Refactoring of Java Client to couchb...       1   
12      Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
420134  CBQE-3067: Use better FTS level compaction con...       1   
420185  CBQE-4209: Added the cbas installation and ser...       1   
420805            Adding subquery tests to watson branch.       1   
421737  CBQE-1898 lww tests are covered in py-entbacku...       1   
422442  CBQE-4378 String manipulation in bash incorrec...       1   

             main_author  revision_number  \
4          Emerson Nolan                2   
5          Emerson Nolan                2   
9          Emerson Nolan                2   
10         Emerson Nolan                2   
12         Emerson Nolan                3   
...                  ...              ...   
420134        Ellie Kidd                2   
420185       Alfred Park                1   
420805        Dylan Paul                1   
421737  Frederick Booker                1   
422442    Angelina Mayer                1   

                                            url PR_number_no_patch_number  \
4       http://review.couchbase.org/#/c/12050/2                     12050   
5       http://review.couchbase.org/#/c/12050/2                     12050   
9       http://review.couchbase.org/#/c/12050/2                    

In [23]:
useful_data['comment_length'] = useful_data['Comment'].str.len()

C:\Users\GNR\AppData\Local\Temp\ipykernel_588\2463225354.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  useful_data['comment_length'] = useful_data['Comment'].str.len()


In [24]:
useful_data

PR_number                                            Comment  \
4       12050_rev2_d   I would prefer that you didn't submit thisI w...   
5       12050_rev2_d  You also need to fix the commit message so it ...   
9       12050_rev2_d   I would prefer that you didn't submit this-1 ...   
10      12050_rev2_d  I don't have a huge opinion on it, really.  My...   
12      12050_rev3_d  I think Rags is working on this today too, so ...   
...              ...                                                ...   
420134  80604_rev2_d                       Got it - will remember that.   
420185  80992_rev1_d  MemoryQuota may set in param with default valu...   
420805  82094_rev1_d  This change has already been reviewed by dev f...   
421737  84162_rev1_d  test suite lww in entbackup restore is also re...   
422442  85613_rev1_d      Need to verify it on both, centos and windows   

                  Author                           Date  \
4          Kenny Maynard  2012-01-04 22:01:33.920000000   
5          Kenny Maynard  2012-01-04 22:02:19.848000000   
9          Kenny Maynard  2012-01-05 22:42:21.672000000   
10             Noah Dean  2012-01-06 03:50:33.697000000   
12         Andreas Short  2012-01-08 21:36:57.811000000   
...                  ...                            ...   
420134        Ellie Kidd  2017-07-13 18:41:21.038000000   
420185  Frederick Booker  2017-07-18 15:43:30.472000000   
420805    Monique Barker  2017-08-09 02:25:02.503000000   
421737  Frederick Booker  2017-10-09 19:52:13.747000000   
422442  Frederick Booker  2017-11-17 18:42:57.845000000   

                                           Commit_message  \
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
9       Refactoring changes summary:Membase has been r...   
10      Refactoring changes summary:Membase has been r...   
12      Refactoring changes summary:Membase has been r...   
...                                                   ...   
420134                                                NaN   
420185                                                NaN   
420805                                                NaN   
421737                                                NaN   
422442                                                NaN   

                                             Descriptions  status  \
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
9       Potential Refactoring of Java Client to couchb...       1   
10      Potential Refactoring of Java Client to couchb...       1   
12      Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
420134  CBQE-3067: Use better FTS level compaction con...       1   
420185  CBQE-4209: Added the cbas installation and ser...       1   
420805            Adding subquery tests to watson branch.       1   
421737  CBQE-1898 lww tests are covered in py-entbacku...       1   
422442  CBQE-4378 String manipulation in bash incorrec...       1   

             main_author  revision_number  \
4          Emerson Nolan                2   
5          Emerson Nolan                2   
9          Emerson Nolan                2   
10         Emerson Nolan                2   
12         Emerson Nolan                3   
...                  ...              ...   
420134        Ellie Kidd                2   
420185       Alfred Park                1   
420805        Dylan Paul                1   
421737  Frederick Booker                1   
422442    Angelina Mayer                1   

                                            url PR_number_no_patch_number  \
4       http://review.couchbase.org/#/c/12050/2                     12050   
5       http://review.couchbase.org/#/c/12050/2                     12050   
9       http://review.couchbase.org/#/c/12050/2                    

Deleting again some bots comments.

In [25]:
bot_comments = ['Patch Set','hudson','cv','jasmin', 'Failed','Started', 'http', 'https']
bot_comments_pattern = '|'.join(bot_comments)
useful_data = useful_data[useful_data.Comment.str.contains(bot_comments_pattern) == False]

In [26]:
useful_data

PR_number                                            Comment  \
4       12050_rev2_d   I would prefer that you didn't submit thisI w...   
5       12050_rev2_d  You also need to fix the commit message so it ...   
9       12050_rev2_d   I would prefer that you didn't submit this-1 ...   
10      12050_rev2_d  I don't have a huge opinion on it, really.  My...   
12      12050_rev3_d  I think Rags is working on this today too, so ...   
...              ...                                                ...   
420134  80604_rev2_d                       Got it - will remember that.   
420185  80992_rev1_d  MemoryQuota may set in param with default valu...   
420805  82094_rev1_d  This change has already been reviewed by dev f...   
421737  84162_rev1_d  test suite lww in entbackup restore is also re...   
422442  85613_rev1_d      Need to verify it on both, centos and windows   

                  Author                           Date  \
4          Kenny Maynard  2012-01-04 22:01:33.920000000   
5          Kenny Maynard  2012-01-04 22:02:19.848000000   
9          Kenny Maynard  2012-01-05 22:42:21.672000000   
10             Noah Dean  2012-01-06 03:50:33.697000000   
12         Andreas Short  2012-01-08 21:36:57.811000000   
...                  ...                            ...   
420134        Ellie Kidd  2017-07-13 18:41:21.038000000   
420185  Frederick Booker  2017-07-18 15:43:30.472000000   
420805    Monique Barker  2017-08-09 02:25:02.503000000   
421737  Frederick Booker  2017-10-09 19:52:13.747000000   
422442  Frederick Booker  2017-11-17 18:42:57.845000000   

                                           Commit_message  \
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
9       Refactoring changes summary:Membase has been r...   
10      Refactoring changes summary:Membase has been r...   
12      Refactoring changes summary:Membase has been r...   
...                                                   ...   
420134                                                NaN   
420185                                                NaN   
420805                                                NaN   
421737                                                NaN   
422442                                                NaN   

                                             Descriptions  status  \
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
9       Potential Refactoring of Java Client to couchb...       1   
10      Potential Refactoring of Java Client to couchb...       1   
12      Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
420134  CBQE-3067: Use better FTS level compaction con...       1   
420185  CBQE-4209: Added the cbas installation and ser...       1   
420805            Adding subquery tests to watson branch.       1   
421737  CBQE-1898 lww tests are covered in py-entbacku...       1   
422442  CBQE-4378 String manipulation in bash incorrec...       1   

             main_author  revision_number  \
4          Emerson Nolan                2   
5          Emerson Nolan                2   
9          Emerson Nolan                2   
10         Emerson Nolan                2   
12         Emerson Nolan                3   
...                  ...              ...   
420134        Ellie Kidd                2   
420185       Alfred Park                1   
420805        Dylan Paul                1   
421737  Frederick Booker                1   
422442    Angelina Mayer                1   

                                            url PR_number_no_patch_number  \
4       http://review.couchbase.org/#/c/12050/2                     12050   
5       http://review.couchbase.org/#/c/12050/2                     12050   
9       http://review.couchbase.org/#/c/12050/2                    

--------------------------------------------------------END--------------------------------------------------------------------

# Readability 
The Flesch-Kincaid Grade Level

In [27]:
len(useful_data)

30101

In [28]:
useful_data['Flesch-Kincaid'] = 0

C:\Users\GNR\AppData\Local\Temp\ipykernel_588\436516747.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  useful_data['Flesch-Kincaid'] = 0


In [29]:
import textstat

for n in range(len(useful_data)):
    useful_data['Flesch-Kincaid'].iloc[n] = textstat.flesch_reading_ease(useful_data['Comment'].iloc[n])
#textstat.flesch_reading_ease(text)

C:\Users\GNR\AppData\Local\Temp\ipykernel_588\1877658995.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  useful_data['Flesch-Kincaid'].iloc[n] = textstat.flesch_reading_ease(useful_data['Comment'].iloc[n])


In [32]:
useful_data

PR_number                                            Comment  \
4       12050_rev2_d   I would prefer that you didn't submit thisI w...   
5       12050_rev2_d  You also need to fix the commit message so it ...   
9       12050_rev2_d   I would prefer that you didn't submit this-1 ...   
10      12050_rev2_d  I don't have a huge opinion on it, really.  My...   
12      12050_rev3_d  I think Rags is working on this today too, so ...   
...              ...                                                ...   
420134  80604_rev2_d                       Got it - will remember that.   
420185  80992_rev1_d  MemoryQuota may set in param with default valu...   
420805  82094_rev1_d  This change has already been reviewed by dev f...   
421737  84162_rev1_d  test suite lww in entbackup restore is also re...   
422442  85613_rev1_d      Need to verify it on both, centos and windows   

                  Author                           Date  \
4          Kenny Maynard  2012-01-04 22:01:33.920000000   
5          Kenny Maynard  2012-01-04 22:02:19.848000000   
9          Kenny Maynard  2012-01-05 22:42:21.672000000   
10             Noah Dean  2012-01-06 03:50:33.697000000   
12         Andreas Short  2012-01-08 21:36:57.811000000   
...                  ...                            ...   
420134        Ellie Kidd  2017-07-13 18:41:21.038000000   
420185  Frederick Booker  2017-07-18 15:43:30.472000000   
420805    Monique Barker  2017-08-09 02:25:02.503000000   
421737  Frederick Booker  2017-10-09 19:52:13.747000000   
422442  Frederick Booker  2017-11-17 18:42:57.845000000   

                                           Commit_message  \
4       Refactoring changes summary:Membase has been r...   
5       Refactoring changes summary:Membase has been r...   
9       Refactoring changes summary:Membase has been r...   
10      Refactoring changes summary:Membase has been r...   
12      Refactoring changes summary:Membase has been r...   
...                                                   ...   
420134                                                NaN   
420185                                                NaN   
420805                                                NaN   
421737                                                NaN   
422442                                                NaN   

                                             Descriptions  status  \
4       Potential Refactoring of Java Client to couchb...       1   
5       Potential Refactoring of Java Client to couchb...       1   
9       Potential Refactoring of Java Client to couchb...       1   
10      Potential Refactoring of Java Client to couchb...       1   
12      Potential Refactoring of Java Client to couchb...       1   
...                                                   ...     ...   
420134  CBQE-3067: Use better FTS level compaction con...       1   
420185  CBQE-4209: Added the cbas installation and ser...       1   
420805            Adding subquery tests to watson branch.       1   
421737  CBQE-1898 lww tests are covered in py-entbacku...       1   
422442  CBQE-4378 String manipulation in bash incorrec...       1   

             main_author  revision_number  \
4          Emerson Nolan                2   
5          Emerson Nolan                2   
9          Emerson Nolan                2   
10         Emerson Nolan                2   
12         Emerson Nolan                3   
...                  ...              ...   
420134        Ellie Kidd                2   
420185       Alfred Park                1   
420805        Dylan Paul                1   
421737  Frederick Booker                1   
422442    Angelina Mayer                1   

                                            url PR_number_no_patch_number  \
4       http://review.couchbase.org/#/c/12050/2                     12050   
5       http://review.couchbase.org/#/c/12050/2                     12050   
9       http://review.couchbase.org/#/c/12050/2                    

In [33]:
compression_opts = dict(method='zip',
                        archive_name='all_data_for_analysis.csv') 
useful_data.to_csv('all_data_for_analysis.zip', index=False, compression=compression_opts)